# 深度学习神经网络的Hello World
主要参考于[DeepLearning.ai课程](https://github.com/lmoroney/dlaicourse)

就像每一个应用程序一样，你应该从一些很简单代码开始学习，并且这些代码可以展示出程序如何工作的整体框架。

在学习神经网络时，一个较简单的例子即是学习两类数字之间的关系。

如：我们已经知道这些数字之间的规则，编写出如下代码：

**Question:**
* 那么，如何创建神经网络使它能够找到这个规则，完成等效的工作？

**Answer:**
* 使用数据，通过给神经网络输入一组x和y，使它能够弄清楚数据之间的关系

显然，这是一个可能与之前编写的习惯范例完全不同的范例，因此让我们逐步完成它！

## 1. 导入包

In [2]:
import tensorflow as tf
import numpy as np
from tensorflow import keras

## 2. 定义和编译神经网络
创建一个很简单的神经网络，它只有一层，一个神经元，并且输入的也是一个值

In [3]:
# 实例化一个序列化模型
model = tf.keras.Sequential([keras.layers.Dense(units = 1, input_shape = [1])])

接下来，让我们编译这个神经网络，但是我们必须要明确两个函数：损失函数和优化函数

我们人类可以很轻松的得知这些数之间的关系是y=2x-1，但是计算机没那么简单。

**loss：**当计算机尝试“学习”时，会做出猜测……也许是y = 10x + 10。 损失函数可以根据已知的正确答案来衡量猜测的答案，并衡量其执行的好坏程度。

**optimizer：**然后使用优化器函数进行另一次猜测。根据损失函数的运行情况，它会尽量减少损失。它可能会得到像y=5x+5这样的结果，虽然仍然很糟糕，但更接近正确的结果(即损失更低)

神经网络将重复此操作，我们让它使用“均方误差”来表示损失，并使用“随机梯度下降”来表示优化器的方法。 (您还不需要了解它们的数学原理，但是您可以看到它们起作用了！)

In [4]:
model.compile(optimizer='sgd',loss='mean_squared_error')

随着不断学习，将了解针对不同情况的不同且适当的损失和优化器功能。

## 3. 导入数据
接下来，将输入一些数据给神经网络。 

In [5]:
xs = np.array([-1.0,  0.0, 1.0, 2.0, 3.0, 4.0], dtype=float)
ys = np.array([-3.0, -1.0, 1.0, 3.0, 5.0, 7.0], dtype=float)

## 4. 训练神经网络
训练神经网络的过程是在model.fit调用中“学习” Xs和Ys之间的关系。 这就是我们上面讨论过的循环的地方，将对关系进行猜测，测量其好坏（也称为损失），使用优化器进行另一次猜测等。它将针对您所指定的循环周期数进行运行。运行此代码时，可以在右侧看到损失。

In [6]:
model.fit(xs,ys,epochs=500)

Train on 6 samples
Epoch 1/500
6/6 [==============================] - 0s 48ms/sample - loss: 12.8001
Epoch 2/500
6/6 [==============================] - 0s 500us/sample - loss: 10.2983
Epoch 3/500
6/6 [==============================] - 0s 500us/sample - loss: 8.3253
Epoch 4/500
6/6 [==============================] - 0s 2ms/sample - loss: 6.7685
Epoch 5/500
6/6 [==============================] - 0s 333us/sample - loss: 5.5391
Epoch 6/500
6/6 [==============================] - 0s 667us/sample - loss: 4.5675
Epoch 7/500
6/6 [==============================] - 0s 3ms/sample - loss: 3.7989
Epoch 8/500
6/6 [==============================] - 0s 167us/sample - loss: 3.1899
Epoch 9/500
6/6 [==============================] - 0s 500us/sample - loss: 2.7066
Epoch 10/500
6/6 [==============================] - 0s 500us/sample - loss: 2.3224
Epoch 11/500
6/6 [==============================] - 0s 667us/sample - loss: 2.0161
Epoch 12/500
6/6 [==============================] - 0s 1ms/sample - loss: 1.7713

Epoch 101/500
6/6 [==============================] - 0s 1ms/sample - loss: 0.1483
Epoch 102/500
6/6 [==============================] - 0s 333us/sample - loss: 0.1453
Epoch 103/500
6/6 [==============================] - 0s 500us/sample - loss: 0.1423
Epoch 104/500
6/6 [==============================] - 0s 333us/sample - loss: 0.1394
Epoch 105/500
6/6 [==============================] - 0s 333us/sample - loss: 0.1365
Epoch 106/500
6/6 [==============================] - 0s 333us/sample - loss: 0.1337
Epoch 107/500
6/6 [==============================] - 0s 1ms/sample - loss: 0.1310
Epoch 108/500
6/6 [==============================] - 0s 667us/sample - loss: 0.1283
Epoch 109/500
6/6 [==============================] - 0s 1ms/sample - loss: 0.1256
Epoch 110/500
6/6 [==============================] - 0s 333us/sample - loss: 0.1231
Epoch 111/500
6/6 [==============================] - 0s 333us/sample - loss: 0.1205
Epoch 112/500
6/6 [==============================] - 0s 1ms/sample - loss: 0.1181


Epoch 200/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0190
Epoch 201/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0186
Epoch 202/500
6/6 [==============================] - 0s 333us/sample - loss: 0.0182
Epoch 203/500
6/6 [==============================] - 0s 2ms/sample - loss: 0.0179
Epoch 204/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0175
Epoch 205/500
6/6 [==============================] - 0s 1ms/sample - loss: 0.0171
Epoch 206/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0168
Epoch 207/500
6/6 [==============================] - 0s 333us/sample - loss: 0.0164
Epoch 208/500
6/6 [==============================] - 0s 1ms/sample - loss: 0.0161
Epoch 209/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0158
Epoch 210/500
6/6 [==============================] - 0s 333us/sample - loss: 0.0154
Epoch 211/500
6/6 [==============================] - 0s 833us/sample - loss: 0.015

Epoch 299/500
6/6 [==============================] - 0s 333us/sample - loss: 0.0024
Epoch 300/500
6/6 [==============================] - 0s 667us/sample - loss: 0.0024
Epoch 301/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0023
Epoch 302/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0023
Epoch 303/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0022
Epoch 304/500
6/6 [==============================] - 0s 2ms/sample - loss: 0.0022
Epoch 305/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0022
Epoch 306/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0021
Epoch 307/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0021
Epoch 308/500
6/6 [==============================] - 0s 167us/sample - loss: 0.0020
Epoch 309/500
6/6 [==============================] - 0s 2ms/sample - loss: 0.0020
Epoch 310/500
6/6 [==============================] - 0s 500us/sample - loss: 0.0

Epoch 395/500
6/6 [==============================] - 0s 333us/sample - loss: 3.3210e-04
Epoch 396/500
6/6 [==============================] - 0s 833us/sample - loss: 3.2528e-04
Epoch 397/500
6/6 [==============================] - 0s 500us/sample - loss: 3.1860e-04
Epoch 398/500
6/6 [==============================] - 0s 333us/sample - loss: 3.1205e-04
Epoch 399/500
6/6 [==============================] - 0s 333us/sample - loss: 3.0564e-04
Epoch 400/500
6/6 [==============================] - 0s 500us/sample - loss: 2.9936e-04
Epoch 401/500
6/6 [==============================] - 0s 333us/sample - loss: 2.9321e-04
Epoch 402/500
6/6 [==============================] - 0s 833us/sample - loss: 2.8719e-04
Epoch 403/500
6/6 [==============================] - 0s 500us/sample - loss: 2.8129e-04
Epoch 404/500
6/6 [==============================] - 0s 167us/sample - loss: 2.7551e-04
Epoch 405/500
6/6 [==============================] - 0s 667us/sample - loss: 2.6985e-04
Epoch 406/500
6/6 [=============

Epoch 489/500
6/6 [==============================] - 0s 167us/sample - loss: 4.7205e-05
Epoch 490/500
6/6 [==============================] - 0s 333us/sample - loss: 4.6236e-05
Epoch 491/500
6/6 [==============================] - 0s 1ms/sample - loss: 4.5287e-05
Epoch 492/500
6/6 [==============================] - 0s 333us/sample - loss: 4.4356e-05
Epoch 493/500
6/6 [==============================] - 0s 333us/sample - loss: 4.3445e-05
Epoch 494/500
6/6 [==============================] - 0s 1ms/sample - loss: 4.2552e-05
Epoch 495/500
6/6 [==============================] - 0s 333us/sample - loss: 4.1678e-05
Epoch 496/500
6/6 [==============================] - 0s 333us/sample - loss: 4.0823e-05
Epoch 497/500
6/6 [==============================] - 0s 1ms/sample - loss: 3.9983e-05
Epoch 498/500
6/6 [==============================] - 0s 500us/sample - loss: 3.9163e-05
Epoch 499/500
6/6 [==============================] - 0s 333us/sample - loss: 3.8358e-05
Epoch 500/500
6/6 [===================

现在有了一个经过训练的模型来学习X和y之间的关系，你可以使用这个模型。来预测Y的值

例如：如果X = 10，你认为Y会是多少?

In [7]:
print(model.predict([10.0]))

[[18.982117]]


**Question：**
 * 你也许会认为输出的是19，但是结果比这小一点，是18.982117，为什么会这样呢？

**Answer：**
 * 请记住，神经网络对概率进行预测，因此，提供给神经网络的数据，它计算出X和Y之间的关系为Y = 2X-1的可能性非常高，但是只有6个数据点，我们不能确定Y=2X-1一定成立，所以预测的结果非常接近19，但不一定是19。
 
当您使用神经网络时，您将看到这种模式重复出现。你几乎总是处理概率,不确定性,并将做一些编码算出结果是基于概率的,特别是在分类过程中。

